File: src/main/java/eu/karelhovorka/zpevnik/text/SongDisplaySettings.java
package eu.karelhovorka.zpevnik.text;


import eu.karelhovorka.zpevnik.music.Interval;
import eu.karelhovorka.zpevnik.util.Tone;

import static eu.karelhovorka.zpevnik.util.Preconditions.checkNotNull;

public class SongDisplaySettings {

    public static final String LIGHT_THEME = "light";

    public static final String DARK_THEME = "dark";

    public static final SongDisplaySettings DEFAULT = new SongDisplaySettings(true, true, false, false, false, false, Interval.PERFECT_UNISON, Tone.CountryCategory.EASTERN, Tone.ModificationAbbreviation.SHARP, DARK_THEME);

    private boolean displayText;

    private boolean displayChords;

    private boolean hideIdenticalSequences;

    private boolean useBold;

    private boolean doubleColumn;

    private boolean resizeChords;

    private Interval interval;

    private Tone.CountryCategory countryCategory;

    private Tone.ModificationAbbreviation modificationAbbreviation;

    private String theme;

    public SongDisplaySettings(boolean displayText, boolean displayChords, boolean hideIdenticalSequences, boolean useBold, boolean doubleColumn, boolean resizeChords, Interval interval, Tone.CountryCategory countryCategory, Tone.ModificationAbbreviation modificationAbbreviation, String theme) {
        checkNotNull(countryCategory, "countryCategory");
        checkNotNull(modificationAbbreviation, "modificationAbbreviation");
        checkNotNull(theme, "theme");
        checkNotNull(interval, "interval");
        this.displayText = displayText;
        this.doubleColumn = doubleColumn;
        this.displayChords = displayChords;
        this.useBold = useBold;
        this.resizeChords = resizeChords;
        this.hideIdenticalSequences = hideIdenticalSequences;
        this.interval = interval;
        this.countryCategory = countryCategory;
        this.modificationAbbreviation = modificationAbbreviation;
        this.theme = theme;
    }

    public boolean isDisplayText() {
        return displayText;
    }

    public boolean isDisplayChords() {
        return displayChords;
    }

    public boolean isHideIdenticalSequences() {
        return hideIdenticalSequences;
    }

    public Interval getInterval() {
        return interval;
    }

    public Tone.CountryCategory getCountryCategory() {
        return countryCategory;
    }

    public Tone.ModificationAbbreviation getModificationAbbreviation() {
        return modificationAbbreviation;
    }

    public boolean isUseBold() {
        return useBold;
    }

    public boolean isDoubleColumn() {
        return doubleColumn;
    }

    public boolean isResizeChords() {
        return resizeChords;
    }

    public String getTheme() {
        return theme;
    }

    public String getCss() {
        StringBuilder sb = new StringBuilder(theme);
        if (useBold) {
            sb.append(" use-bold");
        }
        if (resizeChords) {
            sb.append(" resize-chords");
        }
        if (doubleColumn) {
            sb.append(" double-column");
        }
        if (hideIdenticalSequences) {
            sb.append(" hide-identical-sequences");
        }
        if (displayChords) {
            sb.append(" display-chords");
        }
        if (displayText) {
            sb.append(" display-text");
        }
        return sb.toString();
    }
}

File: src/commonMain/kotlin/eu/karelhovorka/zpevnik/text/SongDisplaySettings.kt
package eu.karelhovorka.zpevnik.text


import eu.karelhovorka.zpevnik.music.Interval
import eu.karelhovorka.zpevnik.util.Preconditions.checkNotNull
import eu.karelhovorka.zpevnik.util.Tone
import kotlin.jvm.JvmField
import kotlin.jvm.JvmOverloads


data class SongDisplaySettings @JvmOverloads constructor(
    val isDisplayText: Boolean = true,
    val isDisplayChords: Boolean = true,
    val isHideIdenticalSequences: Boolean = false,
    val isUseBold: Boolean = false,
    val isDoubleColumn: Boolean = false,
    val isResizeChords: Boolean = false,
    val showComments: Boolean = false,
    val interval: Interval = Interval.of(0),
    val countryCategory: Tone.CountryCategory = Tone.CountryCategory.EASTERN,
    val modificationAbbreviation: Tone.ModificationAbbreviation = Tone.ModificationAbbreviation.SHARP,
    val theme: String = LIGHT_THEME,
    val fontFamily: String = DEFAULT_FONT
) {

    val css: String
        get() {
            val sb = StringBuilder(theme)
            sb.append(" font-$fontFamily")
            if (isUseBold) {
                sb.append(" use-bold")
            }
            if (isResizeChords) {
                sb.append(" resize-chords")
            }
            if (isDoubleColumn) {
                sb.append(" double-column")
            }
            if (isHideIdenticalSequences) {
                sb.append(" hide-identical-sequences")
            }
            if (isDisplayChords) {
                sb.append(" display-chords")
            }
            if (isDisplayText) {
                sb.append(" display-text")
            }
            if (showComments) {
                sb.append(" show-comments")
            } else {
                sb.append(" hide-comments")
            }
            return sb.toString()
        }

    init {
        checkNotNull(countryCategory, "countryCategory")
        checkNotNull(modificationAbbreviation, "modificationAbbreviation")
        checkNotNull(theme, "theme")
        checkNotNull(interval, "interval")
    }

    companion object {

        @JvmField
        val DEFAULT_FONT = "normal"

        @JvmField
        val LIGHT_THEME = "light"

        @JvmField
        val DARK_THEME = "dark"

        val SYSTEM_THEME = "system"

        @JvmField
        val PRINT_THEME = "print"

        @JvmField
        val LEGACY_THEME = "legacy"

        @JvmField
        val DEFAULT = SongDisplaySettings()
    }
}
